### PR TITLE
Add PKGBUILD file

### PR DESCRIPTION
**Pull request by hodasemi**
Hi,

here is the PKGBUILD file for the XR driver separated. I commited an update for the breezy-gnome file to depend on this.

### CODE REVIEW

File: PKGBUILD
Comment (wheaney) [inline PKGBUILD:30]: This is missing the scripts in `bin`

**Review by wheaney — CHANGES_REQUESTED** [review]
Needs to install `bin` scripts

File: PKGBUILD
Comment (hodasemi) [reply PKGBUILD:30]: Which part is required exactly? The driver runs without crashing on my system in the current configuration.

I still have struggles with breezy-desktop, thus I could not interact with it.

Comment (wheaney) [reply PKGBUILD:30]: All these scripts are necessary: https://github.com/wheaney/XRLinuxDriver/blob/main/bin/setup#L41-L58

Comment (hodasemi) [reply PKGBUILD:30]: I changed how the systemd service is run. In your script you change placeholder with the users home directory. That can all be omitted if you run the script as a user unit (`systemctl --user start xreal-air-driver`). All that is necessary for it to copy the script into `systemd/user/` instead of `systemd/system/`. With `sed` I changed the relevant part that the resulting file looks like:

```
[Unit]
Description=XREAL Air user-space driver
After=network.target

[Service]
Type=simple
ExecStart=xrealAirLinuxDriver
Restart=on-failure

[Install]
WantedBy=multi-user.target
```

I installed the `xrealAirLinuxDriver` binary in the system `/usr/bin` which is by default in `$PATH` thus no path is necessary to use the binary here anymore. The uninstall script makes no sense in a system wide install because on remove it won't be executed, left are only some dot files as with any program you use. 

Regarding the google analytics part: Honestly I don't think it is the right place to use, but who am I to judge. Nevertheless if the installer ever comes to an official repository the script is run once every version and the resulting package is then distributed rendering the execution on build time quite useless, even statistically.

**Comment by hodasemi** [comment]
Is there a reason why you build your own curl version? Why not use the version distributed on the system. With actual build files we can add it as system dependency.

File: PKGBUILD
Comment (wheaney) [reply PKGBUILD:30]: It's already been my intention to move this to a user service but it can't be done without also adding udev rules for all supported devices, so IMO that's another job for another PR if it's not essential for this one.

I don't need the Google Analytics in package manager installs and you're right that uninstall is unnecessary, I was just referring to the xreal_driver_config and verify scripts.

**Comment by wheaney** [comment]
I only added the `curl` static link recently because someone had to install it manually, and since I don't have a great way to handle dependencies with the way I currently distribute it, it was easier just to build it in. So let's leave it as-is for now since it's doing no harm, and once I can switch my setup instructions to just package managers, then remove it.

File: PKGBUILD
Comment (hodasemi) [reply PKGBUILD:30]: The problem is that you can't say which user to add in the service file at build time and at installation time the package is fixed.

Comment (hodasemi) [reply PKGBUILD:30]: Making the udev files quite essentially to make this MR work.

**Comment by hodasemi** [comment]
Makes sense. My cmake-magic is quite limited. How much effort would it make to make that a compile time switch via a variable?

File: PKGBUILD
Comment (hodasemi) [reply PKGBUILD:30]: I added the config script, but the verify script doesn't make much sense, since the executable is in another location the uninstall script is omitted.

**Comment by wheaney** [comment]
> How much effort would it make to make that a compile time switch via a variable?

Probably not much effort at all.

**Comment by wheaney** [comment]
So how does this get tested? What kind of testing have you done? The move-from-root piece has me the most nervous because of the wide range of devices to test, and the potential to break the steam deck integration. One thing we may need to fix before this can go out that you may not have noticed in your testing is whether the device_license integration is working. Specifically, we'll want to move the license file to a directory owned by the user since I *think* right now it'll be owned by root, at least it is on my system. [This issue](https://github.com/wheaney/breezy-desktop/issues/4) describes where I'm planning on moving all files to conform with XDG standards as this becomes a user-service.

**Comment by hodasemi** [comment]
About the points:

* Testing: This is where it actually hangs now. I don't know how :sweat: 
* With this pkgfile everything about the current installation process can be in place as is (this is a big can). If we actually get this to run as a user service things can be changed. If the changes don't work for the steam deck, things can be in place as they are.
* Licensing: I need to put the licenses in place, as I haven't touched that at all. With the pkg file it is not possible to put them into a user directory, since we can have that case that packages can be installed by different user than the user that uses this driver. Also there is not really a way of knowing because the installation has to be done as root anyway.

**Comment by wheaney** [comment]
Just to clarify, the installer doesn't touch license files, they just get downloaded by the driver itself, but since the driver isn't running as root, it can no longer use its current location in `/var/lib` (I think).

**Comment by wheaney** [comment]
Testing would be as simple as:
1. Install the driver from clean slate, verify it runs and downloads/creates the license file
2. Enable and put in mouse mode, verify a supported device moves the mouse
3. Verify /dev/shm/xr_driver_state is created and modified with a live heartbeat and device info
4. Verify config and control flags are abided

To test the IPC sending of the IMU data to external apps (e.g. vulkan layer or gnome extension), easiest would be to install breezy-vulkan and then uninstall the driver (breezy installs a locked verison of the driver) and re-install with this one. Then we can put in virtual display mode and test that the vulkan shader gets this data.

**Comment by hodasemi** [comment]
> Just to clarify, the installer doesn't touch license files, they just get downloaded by the driver itself, but since the driver isn't running as root, it can no longer use its current location in `/var/lib` (I think).

Ah, then we need to put the licenses somewhere in the home directory.

**Comment by wheaney** [comment]
Yeah, that's where [this issue](https://github.com/wheaney/breezy-desktop/issues/4) came in.

**Comment by hodasemi** [comment]
> Testing would be as simple as:
> 
>     1. Install the driver from clean slate, verify it runs and downloads/creates the license file
> 
>     2. Enable and put in mouse mode, verify a supported device moves the mouse
> 
>     3. Verify /dev/shm/xr_driver_state is created and modified with a live heartbeat and device info
> 
>     4. Verify config and control flags are abided
> 
> 
> To test the IPC sending of the IMU data to external apps (e.g. vulkan layer or gnome extension), easiest would be to install breezy-vulkan and then uninstall the driver (breezy installs a locked verison of the driver) and re-install with this one. Then we can put in virtual display mode and test that the vulkan shader gets this data.

I can verify that it runs and the mouse mode does indeed work. The license file however could not be downloaded (as expected actually). `/dev/shm/xr_driver_state` is also created. 
I would say, the first round of successes comes in. I need to figure out the rest now.

**Comment by hodasemi** [comment]
`BreezyDesktop` does indeed successfully communicate with the driver. It successfully downloaded the license file (after manually changing access rights of /var/lib/xr_driver) and I can see the ui now.
If I enable it I can see a calibrating watermark above the desktop on my glasses, but that never goes away. Am I missing something?

**Comment by wheaney** [comment]
Huh, probably something subtle going on if the banner never goes away.

**Comment by wheaney** [comment]
But that's really promising that everything else seems to be working.

**Comment by hodasemi** [comment]
Your current setup regarding `curl` has an error: It links to the local shared library instead of pulling in the static one. If you decide to add an build option to omit a local `curl` build and use a system dependency that would fix it or you link a static build of curl into your executable

```
ldd xrealAirLinuxDriver
..
	libcurl.so.4 => /home/michael/Dokumente/test/src/XRLinuxDriver/build/modules/curl/lib/libcurl.so.4 (0x00007fe68e7f9000)
..
```

**Comment by wheaney** [comment]
Heads up: I merged in a pretty big change today for RayNeo device support. One thing we'll need to resolve is that it needs to install a shared library (.so) file and the driver needs to have the necessary path to it in the `LD_LIBRARY_PATH`.

**Comment by hodasemi** [comment]
Hi, I'm going to merge your changes first and incorporate them into the pkg file. The general instruction to test this is:

1) Download the `PKGBUILD` file and place it somewhere (I recommend outside of the repo f you clone it)
2) Until it is merged you need to change the `url` from `wheaney` to `hodasemi`, because the changes are obviously not available on your master currently
3) execute `makepkg -si` (`s` installs dependencies, `i` installs the package after building it) in the directory of your `PKGBUILD`
4) enable the service: `systemctl --user enable --now xreal-air-driver`

Yesterday I made the required changes for the license file to be downloaded to the requested location in your issue. The download works, I still need to get `Breezy-Desktop` work with the new location.

**Comment by hodasemi** [comment]
It works now.

File: CMakeLists.txt
Comment (wheaney) [inline CMakeLists.txt:139]: Inconsistent indentation

File: src/plugins/breezy_desktop.c
Comment (wheaney) [inline src/plugins/breezy_desktop.c:165]: Let's put `-2` as a `#define` since it's repeated in a few places and has special meaning.

File: src/plugins/device_license.c
Comment (wheaney) [inline src/plugins/device_license.c:24]: No need to duplicate `/xr_driver` here, just append `DEVICE_LICENSE_DIR`

File: src/plugins/device_license.c
Comment (wheaney) [inline src/plugins/device_license.c:21]: A little more context in the naming would be nice, e.g. `XDG_FALLBACK_DIR` or something similar

File: src/plugins/device_license.c
Comment (wheaney) [inline src/plugins/device_license.c:201]: Please move this to `strings.h`

File: PKGBUILD
Comment (wheaney) [inline PKGBUILD:3]: Let's give this a name that's specific to BreezyGNOME for now (e.g. `XRLinuxDriver-BreezyGNOME`). In the future I intend to have the plugins installed as separate dependencies, like how GStreamer has a base install and then separate plugin installs, so I would want `XRLinuxDriver` to be the base installation. In the interim, some way to distinguish this package from that base package I think would be nice.

**Review by wheaney — CHANGES_REQUESTED** [review]
This looks great. Thanks again for the legwork on this. My feedback is mostly nitpicks.

File: CMakeLists.txt
Comment (wheaney) [reply CMakeLists.txt:139]: BTW, prior to this project I wasn't a C developer and I don't know good formatting practices. If there's a good standard formatting that new devs to the project could easily adapt to in their IDE/whatever and there are good linting options for it, I'd be happy to adopt something like that. I've just been flying by the seat of my pants regarding formatting, honestly. No need to incorporate into this PR, we can talk offline if you have opinions here.

Comment (hodasemi) [reply CMakeLists.txt:139]: Strange, locally it looks correctly indented. Maybe we are using different types (tabs, tab-length or whitespace).

I'm no c-dev myself. The project I worked on were c++ and then I used meson as build tool. But as far as I have seen, everyone does whatever he wants regarding formatting in c. There is no rule.

**Comment by wheaney** [comment]
One thing I mentioned after the big rayneo rebase that doesn't look like it was addressed is the changes to the service definition to point to the proper `LD_LIBRARY_PATH`: https://github.com/wheaney/XRLinuxDriver/blob/f45b8fbaa7ef0b3f824a3452a91dc74bffb6ea38/systemd/xreal-air-driver.service#L8

File: CMakeLists.txt
Comment (wheaney) [reply CMakeLists.txt:139]: You should be able to just copy-paste the whitespace leading up to my other libraries included here.

**Review by wheaney — CHANGES_REQUESTED** [review]
Just waiting on the whitespace fix and confirmation that the rayneo lib is being loaded.

**Comment by hodasemi** [comment]
> One thing I mentioned after the big rayneo rebase that doesn't look like it was addressed is the changes to the service definition to point to the proper `LD_LIBRARY_PATH`:
> 
> https://github.com/wheaney/XRLinuxDriver/blob/f45b8fbaa7ef0b3f824a3452a91dc74bffb6ea38/systemd/xreal-air-driver.service#L8

I'm placing the [rayneo lib](https://github.com/hodasemi/XRLinuxDriver/blob/ee5a682acf0c39c39987cab659b21e3fba22a21f/PKGBUILD#L38) into `/usr/lib` and as you may know, it is by default in `PATH`.

**Comment by hodasemi** [comment]
I used `nano` to make the indentations now. It looks way off in `vscode` now. :shrug:

**Comment by wheaney** [comment]
> it is by default in PATH.

Ah, beautiful! And I see you're removing the `Environment` line I linked with your `sed` usage of the `d` flag.

Okay, so this looks ready to go. It's just waiting on me to test. Thanks!

**Review by wheaney — APPROVED** [review]

**Comment by kelvie** [comment]
Why is this a PR to this repo? shouldn't this be in the AUR?

**Comment by hodasemi** [comment]
There were a few more changes required to make it work. I think it is a good first step. Then the file can be submitted to the AUR.

**Comment by wheaney** [comment]
Just tried this today. I think we're missing a dependency:

> Iconv is not found, make sure to provide it in your build environment

**Comment by wheaney** [comment]
I added `libiconv` to my depends and it still failed. One message I'm seeing at the end of the build:

> Could NOT find Iconv (missing: Iconv_INCLUDE_DIR)

**Comment by hodasemi** [comment]
`iconv` is part of glibc on my system:

```
$ pacman -F iconv                                          
core/glibc 2.39+r52+gf8e4623421-1 [Installiert]
    usr/bin/iconv
```

**Comment by wheaney** [comment]
Dunno what the issue was but I probably shouldn't be trying this on Steam Deck anyway :joy:.

So I switched to Manjaro on my laptop and the build and install worked. But now I hit a segfault, it seems like it's during device_license parsing.

**Comment by hodasemi** [comment]
Steam deck isn't really the target for this one, because the package needs to be reinstalled after every system update.

Strange but hard to debug. The local directory `~/.local/state/xr_driver` is created and the `device_license` should be in there.

**Comment by wheaney** [comment]
Alright, so strangely I rebooted in Manjaro today to see if the license file had been put there, and got past that part, so... something we'll need to look into for fresh installs I guess. But I run into another issue:

```bash
Driver has been re-enabled
Using hardware id 611f47e3bc636ef477b0a2a958540143c7b5790175c1376f0bdd98e065f08ff4
Feature productivity_basic granted.
Starting up XR driver
Waiting for glasses
Found device with vendor ID 0x1bbb and product ID 0xaf50
Found device with vendor ID 0x1bbb and product ID 0xaf50
Feature productivity_basic granted.
Device connected, redirecting input to mouse...
libevdev.libevdev_uinput_create_from_device: Permission denied
```

With the `uaccess` flag in udev, I wouldn't expect there to be permission issues.

**Comment by hodasemi** [comment]
I just checked which pid and vid this device stands for. You are using the rayneo. If you compare the [rayneo udev file](https://github.com/hodasemi/XRLinuxDriver/blob/main/udev/70-rayneo-xr.rules) with the [viture](https://github.com/hodasemi/XRLinuxDriver/blob/main/udev/70-rayneo-xr.rules) one, you can see there are sub systems missing. This is kind of on purpose, because I could not find any resource on that information and I don't own such a device.
Thus indeed, that error is kind of expected.

**Comment by wheaney** [comment]
I got this with my XREALs too

**Comment by hodasemi** [comment]
For xreals I just used the file you linked: https://github.com/wheaney/XRLinuxDriver/issues/61#issuecomment-2155514949

With the difference that the `plugdev` group is replaced by `uaccess` and corrects right. `plugdev` was an older approach with the downside that the user had manually be added to this group. As with the rayneos, I don't xreals too, thus there is no way for me to test it. I just assumed that this udev rule was tested by others. Nevertheless, it is probably a good base for someone to start testing which subsystems are used and fix it accordingly.

**Comment by wheaney** [comment]
I don't think the udev file needs all those subsystems. Here's the XREAL udev I used to use before I moved to a systemd service (root access meant I no longer needed udev), and this worked fine: https://github.com/wheaney/XRLinuxDriver/blob/1b1d644d1f2e97f1bcc3a7dd870ef1f90f6d721f/udev/60-xreal-air.rules

I have a feeling there's something else going on here.

**Comment by hodasemi** [comment]
That would be your part, since I can't test the devices I don't have. I can confirm that the viture one works for me.

**Comment by wheaney** [comment]
Oh how dumb, I even noted in my previous comment:
>  the first line of that file is what allowed me to write mouse/controller movements, not necessary for reading from the device.

And now that I take a closer look at the error, I see that it's actually this very same thing I was talking about: in mouse mode it creates a uinput to write out the mouse movement. So the issue I'm seeing is writing, not reading. We should add one more udev file that adds a rule for writing uinput:

> KERNEL=="uinput", SUBSYSTEM=="misc", TAG+="uaccess", OPTIONS+="static_node=uinput"

**Comment by wheaney** [comment]
One other question I had during the setup:
> systemctl --user enable --now xreal-air-driver

There's no way for the pacman install to do this, I'm assuming? Or would it automatically enable after a restart?

**Comment by hodasemi** [comment]
> KERNEL=="uinput", SUBSYSTEM=="misc", TAG+="uaccess", OPTIONS+="static_node=uinput"

That would apply to xreal right?

---

> There's no way for the pacman install to do this, I'm assuming? Or would it automatically enable after a restart?

No, that is also something I looked for, but it seems that is by design.

**Comment by wheaney** [comment]
> That would apply to xreal right?

It should probably just be in its own generic xr-driver udev file, as it's needed for all glasses.

The udev rule alone didn't solve the issue I was seeing, though. It turns out we also need to make sure the `uinput` module is loaded, this normally requires a call like: `sudo modprobe uinput`, but that won't be doable from PKGBUILD. Instead, according to Claude, we can add this to the PKGBUILD:
> install -Dm644 /dev/null "$pkgdir/usr/lib/modules-load.d/$pkgname.conf"
> echo "uinput" > "$pkgdir/usr/lib/modules-load.d/$pkgname.conf"

With that module installed, I finally got mouse output working from the driver. We should add `post_install` messaging to the PKGBUILD that provides instructions for starting the systemd service, and also loading the udev rules using `udevadm` (e.g. `sudo udevadm control --reload-rules && sudo udevadm trigger`).

**Comment by hodasemi** [comment]
> It should probably just be in its own generic xr-driver udev file, as it's needed for all glasses.

I still need to match product and vendor id, otherwise it would apply to everything else, not just xr glass related.

**Comment by hodasemi** [comment]
I added the `hooks.install` file which is linked in the `PKGBUILD`. There is defined that it `modprobe uinput` after package install und reloads udev rules whenever the package is installed or upgraded. I also added an echo that tells the user to enable the systemd service and how to do that.

**Comment by wheaney** [comment]
> I still need to match product and vendor id, otherwise it would apply to everything else, not just xr glass related.

I don't think this makes sense. The uinput creation is not related to the glasses at all. It's a permission you're giving the user, so user processes can create and modify uinput files.

**Comment by wheaney** [comment]
I just tried out the latest and I didn't see this in the terminal:
> start service after install: systemctl --user enable --now xreal-air-driver.service

So I'm not sure if any of the hook.install stuff happened. Is this just an issue with `makepkg` or should the hook be working for me?

**Comment by hodasemi** [comment]
It works for me. A note: it only shows the message on install not on update, that means if you have it already installed and install it once again it doesn't show up. You have to remove it first and install it anew.

```
$ makepkg
...
[100%] Linking C executable xrealAirLinuxDriver
[100%] Built target xrealAirLinuxDriver
==> Entering fakeroot environment...
==> Starting package()...
==> Tidying install...
  -> Removing libtool files...
  -> Purging unwanted files...
  -> Removing static library files...
  -> Stripping unneeded symbols from binaries and libraries...
  -> Compressing man and info pages...
==> Checking for packaging issues...
==> WARNING: Package contains reference to $srcdir
usr/bin/xrealAirLinuxDriver
==> Creating package "XRLinuxDriver-BreezyGNOME"...
  -> Generating .PKGINFO file...
  -> Generating .BUILDINFO file...
  -> Adding install file...                                                               <-- Important line
  -> Generating .MTREE file...
  -> Compressing package...
==> Leaving fakeroot environment.
==> Finished making: XRLinuxDriver-BreezyGNOME 0.1-1 (Do 18 Jul 2024 05:33:50 CEST)
```

```
# pacman -U XRLinuxDriver-BreezyGNOME-0.1-1-x86_64.pkg.tar.zst
...
(1/1) checking for file conflicts                  [######################] 100%
(1/1) checking available disk space                [######################] 100%
:: Processing package changes...
(1/1) installing XRLinuxDriver-BreezyGNOME         [######################] 100%
start service after install: systemctl --user enable --now xreal-air-driver.service       <-- Important line
:: Running post-transaction hooks...
(1/3) Reloading user manager configuration...
(2/3) Reloading device manager configuration...
(3/3) Arming ConditionNeedsUpdate...

```

**Review by wheaney — APPROVED** [review]

**Pull request merged**
